<a href="https://colab.research.google.com/github/ShanChartrand/NECDataCleaning/blob/main/NEC_data_cleaning_May_24.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Importing file

In [2]:
!pip install pandas
import pandas as pd
df = pd.read_csv('https://raw.githubusercontent.com/ShanChartrand/NECDataCleaning/main/necoriginal.csv')


Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


Importing libraries

In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import io
import requests

Viewing original number of rows and columns

In [4]:
# Count the number of rows and columns
num_rows = df.shape[0]
num_columns = df.shape[1]

# Print the results
print("Number of rows:", num_rows)
print("Number of columns:", num_columns)

Number of rows: 19928
Number of columns: 31


Removing rows 2-8064, before metrics such as Jitter, Latency, and Packet loss were reliably measured

In [5]:
# Read the CSV file
df = pd.read_csv('https://raw.githubusercontent.com/ShanChartrand/NECDataCleaning/main/necoriginal.csv')

# Remove rows 2 through 8064
df.drop(df.index[1:8064], inplace=True)

# Save the updated DataFrame 
df.to_csv('https://raw.githubusercontent.com/ShanChartrand/NECDataCleaning/main/necoriginal.csv', index=False)

Counting the new number of rows & columns after removing rows 2-8064

In [6]:
# Count the number of rows and columns
num_rows = df.shape[0]
num_columns = df.shape[1]

# Print the results
print("Number of rows:", num_rows)
print("Number of columns:", num_columns)

Number of rows: 11865
Number of columns: 31


Looking at the % of missing items in columns

In [7]:
# Calculate the percentage of missing cells in each column
missing_percentages = df.isna().sum() / len(df) * 100

# Print the result
print(missing_percentages)

TEST ID                       0.000000
TEST DATE / UTC HOUR          0.000000
GEO LATITUDE                  0.000000
GEO LONGITUDE                 0.000000
DOWNLOAD SPEED                0.000000
UPLOAD SPEED                  0.000000
JITTER                        0.008428
LATENCY                       0.008428
PACKET LOSS                   0.008428
MOBILE TEST                   0.000000
ISP                           0.000000
CLIENT IP                     0.000000
CLIENT CITY                   0.000000
CLIENT PROVINCE               0.000000
CLIENT POSTAL CODE            0.000000
AS NUMBER                     0.008428
IPV6 SUPPORTED                0.000000
DNSSEC SUPPORTED              0.000000
TEST SERVER                   0.000000
browser_test_url              0.008428
CONTRACTED DOWNLOAD SPEED    75.347661
CONTRACTED UPLOAD SPEED      80.311842
CONTRACTED DATA CAP          87.265065
CONTRACTED MONTHLY COST      79.241466
SERVICE TYPE                 73.881163
USER SATISFACTION        

Removing columns with greater than 68% missing values, and then listing the remaining number of rows & columns

In [8]:
# List of columns to remove
columns_to_remove = [
    'CONTRACTED DOWNLOAD SPEED',
    'CONTRACTED UPLOAD SPEED',
    'CONTRACTED DATA CAP',
    'CONTRACTED MONTHLY COST',
    'SERVICE TYPE',
    'USER SATISFACTION',
    'CONNECTION TYPE',
    'USER COMMENTS',
    'HEATMAP_PROJECT_AREA'
]

# Remove the columns
df.drop(columns=columns_to_remove, inplace=True)

# Save the updated DataFrame 
df.to_csv('https://raw.githubusercontent.com/ShanChartrand/NECDataCleaning/main/necoriginal.csv', index=False)

# Count the number of rows and columns
num_rows = df.shape[0]
num_columns = df.shape[1]

# Print the results
print("Number of rows:", num_rows)
print("Number of columns:", num_columns)

Number of rows: 11865
Number of columns: 22


Checking for duplicates in the TEST ID column

In [9]:
# Count the number of duplicated TEST ID values
num_duplicates = df.duplicated(subset='TEST ID', keep=False).sum()

# Print the number of duplicates
print(f"There are {num_duplicates} rows with duplicated TEST ID values.")

There are 0 rows with duplicated TEST ID values.


Looking for duplicate rows

In [10]:
# Find duplicate rows based on all columns
duplicates = df[df.duplicated()]

# Display the duplicate rows (if any)
if len(duplicates) > 0:
    print(f"Found {len(duplicates)} duplicate rows:")
    print(duplicates)
else:
    print("No duplicate rows found.")

No duplicate rows found.


Looking at the % of missing items in columns

In [11]:
# Calculate the percentage of missing cells in each column
missing_percentages = df.isna().sum() / len(df) * 100

# Print the result
print(missing_percentages)

TEST ID                  0.000000
TEST DATE / UTC HOUR     0.000000
GEO LATITUDE             0.000000
GEO LONGITUDE            0.000000
DOWNLOAD SPEED           0.000000
UPLOAD SPEED             0.000000
JITTER                   0.008428
LATENCY                  0.008428
PACKET LOSS              0.008428
MOBILE TEST              0.000000
ISP                      0.000000
CLIENT IP                0.000000
CLIENT CITY              0.000000
CLIENT PROVINCE          0.000000
CLIENT POSTAL CODE       0.000000
AS NUMBER                0.008428
IPV6 SUPPORTED           0.000000
DNSSEC SUPPORTED         0.000000
TEST SERVER              0.000000
browser_test_url         0.008428
HEATMAP_MUNICIPALITY     0.000000
HEATMAP_CENSUS_TRACTS    0.000000
dtype: float64


As per above, we can see that the columns that have missing items each have exactly 0.008428%, which indicates there was a test interruption or failure with CIRA. Will remove those rows then confirm the % of missing items per column.

In [12]:
# List of columns to check for missing values
columns_to_check = ['JITTER', 'LATENCY', 'PACKET LOSS', 'AS NUMBER', 'browser_test_url']

# Remove rows with missing values in specified columns
df.dropna(subset=columns_to_check, inplace=True)

# Save the updated DataFrame 
df.to_csv('https://raw.githubusercontent.com/ShanChartrand/NECDataCleaning/main/necoriginal.csv', index=False)

# Calculate the percentage of missing cells in each column
missing_percentages = df.isna().sum() / len(df) * 100

# Print the result
print(missing_percentages)

TEST ID                  0.0
TEST DATE / UTC HOUR     0.0
GEO LATITUDE             0.0
GEO LONGITUDE            0.0
DOWNLOAD SPEED           0.0
UPLOAD SPEED             0.0
JITTER                   0.0
LATENCY                  0.0
PACKET LOSS              0.0
MOBILE TEST              0.0
ISP                      0.0
CLIENT IP                0.0
CLIENT CITY              0.0
CLIENT PROVINCE          0.0
CLIENT POSTAL CODE       0.0
AS NUMBER                0.0
IPV6 SUPPORTED           0.0
DNSSEC SUPPORTED         0.0
TEST SERVER              0.0
browser_test_url         0.0
HEATMAP_MUNICIPALITY     0.0
HEATMAP_CENSUS_TRACTS    0.0
dtype: float64


We had 11865 rows, let's see how many were removed

In [13]:
# Count the number of rows
num_rows = df.shape[0]

# Print the results
print("Number of rows:", num_rows)

Number of rows: 11863


We can see that only 2 rows were removed. Now to determine the # of unique ISPs

In [14]:
unique_isp_count = len(df['ISP'].unique())
print("Number of unique ISPs:", unique_isp_count)

Number of unique ISPs: 84


Now we must standardize the ISPs and then count the # of unique ISPs after standardizatioin



In [15]:
def standardize_isp(isp):
    replacements = {
        "Access Communications  Co-operative Limited": "Access Communications Co-operative Limited",
        "Amanah Tech": "Amanah Tech Inc.",
        "Amazon.com, Inc.": "Amazon Technologies Inc.",
        "Bell MTS": "Bell Canada",
        "Bell Mobility": "Bell Canada",
        "Bell Mobility, Inc.": "Bell Canada",
        "Bell MTS Data Centres G.P": "Bell Canada",
        "CIK Telecom": "CIK Telecom INC",
        "Commstream Communications": "Commstream Communications Inc",
        "Corridor Communications, INC.": "Corridor Communications",
        "Manitoba First Nations Education Resource Centre Inc.": "Manitoba First Nations Education Resource Centre",
        "Manitoba Hydro International LTD": "Manitoba Hydro",
        "Manitoba Hydro Telecom": "Manitoba Hydro",
        "MTS Allstream": "MTS Inc.",
        "MTS Inc": "MTS Inc.",
        "MTS Allstream Inc.": "MTS Inc.",
        "MTS Data Centres": "MTS Inc.",
        "Red River College of Applied Arts": "Red River College",
        "Red River College of Applied Arts, Science and Tec": "Red River College",
        "Rogers Cable": "Rogers Communications",
        "Rogers Communications Canada Inc.": "Rogers Communications",
        "Full Throttle Networks": "Full Throttle Networks Inc.",
        "Postmedia Network": "Postmedia Network Inc",
        "Radiant Communications Canada Ltd.": "Radiant Communications Canada",
        "Shaw Communications Inc.": "Shaw Communications",
        "Shaw Telecom G.P.": "Shaw Communications",
        "SpaceX Services": "SpaceX",
        "SpaceX Starlink": "SpaceX",
        "TekSavvy Solutions, Inc.": "TekSavvy Solutions",
        "TELUS Communications Inc": "Telus Communications",
        "TELUS Communications Inc.": "Telus Communications",
        "TeraGo Networks Inc": "TeraGo Networks",
        "TeraGo Networks Inc.": "TeraGo Networks",
        "Total Server Solutions L.L.C": "Total Server Solutions L.L.C",
        "Voyageur Internet Inc": "Voyageur Internet",
        "Xplornet Communications Inc": "Xplornet Communications",
        "Xplornet Communications Inc.": "Xplornet Communications"
    }
    
    return replacements.get(isp, isp)

# Example usage:
# Assuming you have a DataFrame named 'df' with a column named 'ISP'
df['ISP'] = df['ISP'].apply(standardize_isp)

unique_isp_count = len(df['ISP'].unique())
print("Number of unique ISPs:", unique_isp_count)

Number of unique ISPs: 73


I'm going to list all the different ISPs

In [16]:
unique_isps = df['ISP'].unique()
for isp in unique_isps:
    print(isp)

Shaw Communications
Bell Canada
WiBand Communications
TekSavvy Solutions
Altima Telecom
Cipherkey Exchange Corp.
Red River College
CIK Telecom INC
Voyageur Internet
Xplornet Communications
MERLIN
The Exchange Global Server Center
Amanah Tech Inc.
Primus Telecommunications Canada Inc.
ViaNetTV Inc
Rogers Communications
City of Winnipeg
Stafford Communications, Inc.
Full Throttle Networks Inc.
MTS Inc.
Total Server Solutions L.L.C.
Questzone.Net, Inc.
Telus Communications
RBC Convention Centre Winnipeg
Cogent Communications
BIGPIPE
University of Manitoba
Corus Entertainment Inc
Postmedia Network Inc
Fastnet Communications
T. Grand Networks Inc.
Netminders Data Solution
Manitoba Hydro
TeraGo Networks
Alberta Health Services
South Winnipeg Technical Centre
Swift High Speed.com
VPN
GloboTech Communications
Corridor Communications
LES.NET
VOI NET INC.
Packet Exchange Limited
Performive LLC
Radiant Communications Canada
SpaceX
Access Communications Co-operative Limited
BH Telecom Corp.
MRNet


Looking at for upload/download speeds of 0 mbps

In [17]:
# Count the number of cells equal to zero in the 'DOWNLOAD SPEED' column
download_speed_zeros = (df['DOWNLOAD SPEED'] == 0).sum()

# Count the number of cells equal to zero in the 'UPLOAD SPEED' column
upload_speed_zeros = (df['UPLOAD SPEED'] == 0).sum()

# Print the counts
print("Number of cells equal to zero in 'DOWNLOAD SPEED' column:", download_speed_zeros)
print("Number of cells equal to zero in 'UPLOAD SPEED' column:", upload_speed_zeros)

Number of cells equal to zero in 'DOWNLOAD SPEED' column: 0
Number of cells equal to zero in 'UPLOAD SPEED' column: 5


I now need to remove the 5 rows that have upload speeds of 0, and then list how many cells = 0 to confirm removal was successful

In [19]:
# Count the number of cells equal to zero in the 'UPLOAD SPEED' column before removal
upload_speed_zeros_before = (df['UPLOAD SPEED'] == 0).sum()

# Remove rows where UPLOAD SPEED is equal to 0
df = df[df['UPLOAD SPEED'] != 0]

# Count the number of cells equal to zero in the 'UPLOAD SPEED' column after removal
upload_speed_zeros_after = (df['UPLOAD SPEED'] == 0).sum()

# Save the updated DataFrame 
df.to_csv('https://raw.githubusercontent.com/ShanChartrand/NECDataCleaning/main/necoriginal.csv', index=False)

# Print the count before and after removal
print("Number of cells equal to zero in 'UPLOAD SPEED' column before removal:", upload_speed_zeros_before)
print("Number of cells equal to zero in 'UPLOAD SPEED' column after removal:", upload_speed_zeros_after)

Number of cells equal to zero in 'UPLOAD SPEED' column before removal: 0
Number of cells equal to zero in 'UPLOAD SPEED' column after removal: 0
